lib/MySQL_Thread.cpp

```diff
@@ -1288,66 +1550,56 @@ void MySQL_Thread::process_all_sessions_connections_handler() {
 	}
 }
 
-void MySQL_Thread::register_session_connection_handler(MySQL_Session *_sess) {
+void MySQL_Thread::register_session_connection_handler(MySQL_Session *_sess, bool _new) {
 	if (mysql_sessions_connections_handler==NULL) return;
-	mysql_sessions_connections_handler->add(_sess);
 	_sess->thread=this;
 	_sess->connections_handler=true;
 	proxy_debug(PROXY_DEBUG_NET,1,"Thread=%p, Session=%p -- Registered new session for connection handler\n", _sess->thread, _sess);
+	if (_new) {
+		mysql_sessions->add(_sess);
+	} else {
+		mysql_sessions_connections_handler->add(_sess);
+	}	
 }
 
-void MySQL_Thread::unregister_session_connection_handler(int idx) {
+void MySQL_Thread::unregister_session_connection_handler(int idx, bool _new) {
 	if (mysql_sessions_connections_handler==NULL) return;
 	proxy_debug(PROXY_DEBUG_NET,1,"Thread=%p, Session=%p -- Unregistered session\n", this, mysql_sessions_connections_handler->index(idx));
-	mysql_sessions_connections_handler->remove_index_fast(idx);
-}
-
-
-void MySQL_Thread::myds_backend_set_failed_connect(MySQL_Data_Stream *myds, unsigned int n) {
-	if (curtime>mypolls.last_recv[n]+10000000) {
-		proxy_error("connect() timeout . curtime: %llu , last_recv: %llu , failed after %lluus . fd: %d , myds_type: %s\n", curtime, mypolls.last_recv[n] , (curtime-mypolls.last_recv[n]) , myds->fd, (myds->myds_type==MYDS_BACKEND_PAUSE_CONNECT ? "MYDS_BACKEND_PAUSE_CONNECT" : "MYDS_BACKEND_NOT_CONNECTED" ) );
-		myds->myds_type=MYDS_BACKEND_FAILED_CONNECT;
-		myds->sess->pause=curtime+10000000;
-		myds->sess->to_process=1;
-	}
-}
-
-void MySQL_Thread::myds_backend_pause_connect(MySQL_Data_Stream *myds) {
-	proxy_error("connect() error on fd %d . Pausing ...\n", myds->fd);
-	myds->myds_type=MYDS_BACKEND_PAUSE_CONNECT;
-	myds->sess->pause=curtime+10000000;
-	myds->sess->to_process=1;
-}
-
-void MySQL_Thread::myds_backend_first_packet_after_connect(MySQL_Data_Stream *myds, unsigned int n) {
-	int optval;
-	socklen_t optlen=sizeof(optval);
-	getsockopt(myds->fd, SOL_SOCKET, SO_ERROR, &optval, &optlen);
-	if (optval==0) {
-		mypolls.last_recv[n]=curtime;
-		myds->myds_type=MYDS_BACKEND;
-		myds->sess->pause=0;
+	if (_new) {
+		mysql_sessions->remove_index_fast(idx);
 	} else {
-		fprintf(stderr,"Connect() error\n");
-		myds->myds_type=MYDS_BACKEND_PAUSE_CONNECT;
-		myds->sess->pause=curtime+10000000;
+		mysql_sessions_connections_handler->remove_index_fast(idx);
 	}
 }
 
+
 void MySQL_Thread::listener_handle_new_connection(MySQL_Data_Stream *myds, unsigned int n) {
 	int c;
-	c=accept(myds->fd, NULL, NULL);
+	struct sockaddr *addr=(struct sockaddr *)malloc(sizeof(struct sockaddr));
+	socklen_t addrlen=sizeof(struct sockaddr);
+	memset(addr, 0, sizeof(struct sockaddr));
+	if (GloMTH->num_threads > 1) {
+		// there are more than 1 thread . We pause for a little bit to avoid all connections to be handled by the same thread
```

Comment:
One of the limitation in proxysql is that a session assigned to a thread will always be executed by that specific thread. This limitation is by design, and it is due to: a) a custom memory allocator and b) each thread calls poll() independently. Since the customer memory allocator is disabled by default, this architecture may change in future if a way to shuffle sessions will be added.
As a consequence of the fact that each thread calls poll() , all threads are all listening on the same port, but when calling accept() only one of them will succeed.

After introducing table stats_mysql_processlist and being able to monitor what queries were being executed inside ProxySQL and from what thread, it turned out that new connections were not being distributed evenly to all threads: some threads were processing more connections than others.
This change tries to solve this problem with incoming connections.
Yet, as sessions may drop during the execution and/or some sessions can run more queries than others or use more resources, there is no guaranteed that a better distribution of new connections leads to better load distribution.
An interesting blog post:

https://blogs.oracle.com/pgdh/entry/how_not_to_load_balance1